src/ProgressOnderwijsUtils/Collections/Tree.cs

```diff
@@ -46,30 +59,67 @@ public static Tree<T> BuildRecursively<T>(T root, IReadOnlyDictionary<T, IReadOn
-        public static Tree<TR> Select<T, TR>(this Tree<T> tree, Func<T, TR> mapper)
-        {
-            var todo = new Stack<Tree<T>>(16);
-            var reconstruct = new Stack<Tree<T>>(16);
-            todo.Push(tree);
-            while (todo.Count > 0) {
-                var next = todo.Pop();
-                reconstruct.Push(next);
-                var children = next.Children;
-                for (var i = children.Count - 1; i >= 0; i--) {
-                    todo.Push(children[i]);
-                }
-            }
+        public static Tree<TR> Select<TTree, T, TR>(this IRecursiveStructure<TTree, T> tree, Func<T, TR> mapper)
+            where TTree : IRecursiveStructure<TTree, T>
+            => tree.TypedThis.Select(node => mapper(node.NodeValue));
 
-            var output = new Stack<Tree<TR>>(16);
-            while (reconstruct.Count > 0) {
-                var next = reconstruct.Pop();
-                var mappedChildren = new Tree<TR>[next.Children.Count];
-                for (var i = 0; i < mappedChildren.Length; i++) {
-                    mappedChildren[i] = output.Pop();
+        /// <summary>
+        /// Builds a copy of this tree with the same structure, but with different node values, as computed by the mapper argument.
+        /// mapper is called bottom-up, in reverse preorder traversal (i.e. children before the node, and the last child first before the first).
+        /// </summary>
+        [Pure]
+        public static Tree<TR> Select<TTree, TR>(this TTree tree, Func<TTree, TR> mapper)
+            where TTree : IRecursiveStructure<TTree>
+            => CachedTreeBuilder<TTree, TR>.Resolve(tree, o => o.Children, (o, kids) => Node(mapper(o), kids));
+
+        /// <summary>
+        /// Recreates a copy of this tree with both structure and node-values altered, as computed by the mapper arguments.
+        /// mapValue and mapStructure are called exactly once for each node in the tree.
+        /// mapValue is passed the old node and should return its new value.
+        /// mapStructure is passed the old node, its new value, and the already-mapped children and should return the nodes that should replace the old one in the tree.
+        ///
+        /// mapValue and mapStructure are called bottom-up, in reverse preorder traversal (i.e. children before the node, and the last child first before the first).
+        /// For a given node, mapValue is called first (and its return value passed to mapStructure).
+        /// </summary>
+        [Pure]
+        public static Tree<TR>[] Rebuild2<TTree, TR>(this TTree tree, Func<TTree, TR> mapValue, Func<TTree, TR, IReadOnlyList<Tree<TR>>, IEnumerable<Tree<TR>?>?> mapStructure)
+            where TTree : IRecursiveStructure<TTree>
+            => CachedTreeBuilder<TTree, Tree<TR>[]>.Resolve(tree, n => n.Children, (n, kids) => Node(mapStructure(n, mapValue(n), kids.SelectMany(k => k.NodeValue).ToArray()).EmptyIfNull().WhereNotNull().ToArray())).NodeValue;

```

Comment:
Dit is ook een rebuild - is dit beter?  Het is wel korter, maar mogelijk een beetje brainbending.

Het is ook veel trager, bringing me to option 3...

Reply:
Als je het zo zegt klinkt het niet heel aantrekkelijk.

Reply:
Ik bedoel: kijk naar de code. Het is *veeeel* korter.  Maar obvious is het niet.

Welke zou jij kiezen?

Reply:
De anderen zijn ook niet obvious en dat is ook niet heel belangrijk, dus waarschijnlijk deze.

Reply:
ok